main/src/ca/uwaterloo/flix/language/phase/Inliner.scala

```diff
@@ -136,13 +136,31 @@ object Inliner {
     case OccurrenceAst.Expression.ApplyClo(exp, args, tpe, purity, loc) =>
       val e = visitExp(exp, subst0)
       val as = args.map(visitExp(_, subst0))
+      e match {
+        case LiftedAst.Expression.Closure(sym, _, _, _) =>
+          val def1 = root.defs.apply(sym)
+          // If `def1` is a single non-self call or
+          // it is trivial
+          // then inline the body of `def1`
+          if (def1.context.isNonSelfCall || isTrivialExp(def1.exp)) {
+            val e1 = convertTailCall(def1.exp)
+            bindFormals(e1, def1.fparams.map(_.sym), as, Map.empty)
+          } else {
+            LiftedAst.Expression.ApplyClo(e, as, tpe, purity, loc)
+          }
+        case _ =>
+          val as = args.map(visitExp(_, subst0))
+          LiftedAst.Expression.ApplyClo(e, as, tpe, purity, loc)
+      }
       LiftedAst.Expression.ApplyClo(e, as, tpe, purity, loc)
```

Comment:
I think there is a mistake, because you always return this line, no matter the previous lines.

Reply:
(Note if the code was written in Flix then you would get a compile error because you have pure computation whose result is discarded...)